From 93282a1c744ea34c7d4dca6c6b918911eae1670e Mon Sep 17 00:00:00 2001
From: Phil Sutter <psutter@redhat.com>
Date: Fri, 17 Jul 2026 11:11:00 +0200
Subject: [PATCH] intervals: do not merge intervals with different timeout

JIRA: https://issues.redhat.com/browse/RHEL-190549
Upstream Status: nftables commit da0bac050c8b2588242727f9915a1ea8bc48ceb2

commit da0bac050c8b2588242727f9915a1ea8bc48ceb2
Author: Pablo Neira Ayuso <pablo@netfilter.org>
Date:   Fri Jan 3 17:40:52 2025 +0100

    intervals: do not merge intervals with different timeout

    If timeout/expiration of contiguous intervals is different, then do not
    merge them.

    Fixes: 81e36530fcac ("src: replace interval segment tree overlap and automerge")
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Phil Sutter <psutter@redhat.com>
---
 src/intervals.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/intervals.c b/src/intervals.c
index 825d9b5..0618b49 100644
--- a/src/intervals.c
+++ b/src/intervals.c
@@ -157,6 +157,8 @@ static void set_prev_elem(struct expr **prev, struct expr *i,
 	mpz_set(prev_range->high, range->high);
 }
 
+static struct expr *interval_expr_key(struct expr *i);
+
 static void setelem_automerge(struct set_automerge_ctx *ctx)
 {
 	struct expr *i, *next, *prev = NULL;
@@ -176,7 +178,9 @@ static void setelem_automerge(struct set_automerge_ctx *ctx)
 		range_expr_value_low(range.low, i);
 		range_expr_value_high(range.high, i);
 
-		if (!prev) {
+		if (!prev ||
+		    interval_expr_key(prev)->timeout != interval_expr_key(i)->timeout ||
+		    interval_expr_key(prev)->expiration != interval_expr_key(i)->expiration) {
 			set_prev_elem(&prev, i, &prev_range, &range);
 			continue;
 		}
